commit 4788069250df4093048d68fac112ce49c9d90363 (refs/changes/65/2065/5)
Author: James Nugent <james@joyent.com>
Date:   2017-06-09T20:03:01-05:00 (2 years, 4 months ago)
    
    OS-6174 Want ability to select admin NIC on boot based on available NICs

diff --git a/overlay/generic/lib/svc/method/net-physical b/overlay/generic/lib/svc/method/net-physical
index 7fbae783..17960dfe 100644
--- a/overlay/generic/lib/svc/method/net-physical
+++ b/overlay/generic/lib/svc/method/net-physical
@@ -498,6 +498,33 @@ if smf_is_globalzone; then
     setup_mtu
 
     # Setup admin NIC
+
+    # If there is no NIC with the admin tag, and the config has
+    # admin_nic_autoselect=true, designate the first NIC reported
+    # by dladm for admin use. This is useful in environments where
+    # the NICs are known to change beneath us.
+    if [[ "${BOOT_smartos}" == "true" ]] && \
+        [[ "${CONFIG_admin_nic_autoselect}" == "true" ]] && \
+        ! nictagadm exists admin ; then
+        autoselected_admin_nic=$(dladm show-phys -m -p -o address | head -n1)
+	if [[ ! -z ${autodetected_admin_nic} ]] ; then
+            echo "ERROR: no NICs found, unable to autoselect admin NIC."
+	    exit ${SMF_EXIT_ERR_CONFIG}
+	fi
+
+	nictagadm add admin "${autoselected_admin_nic}"
+	if [[ $? -ne 0 ]] ; then
+	    echo "ERROR: unable to add admin tag to NIC ${autoselected_admin_nic}"
+	fi
+
+	SYSINFO_NIC_admin=$(dladm show-phys -m -p -o link | head -n1)
+	if [[ ! -z ${SYSINFO_NIC_admin} ]] ; then
+		echo "Autoselected ${SYSINFO_NIC_admin} for use as admin NIC."
+	fi
+
+	nictagadm list
+    fi
+
     if [[ -z "${SYSINFO_NIC_admin}" ]]; then
         echo "ERROR: admin NIC not found, unable to bring up admin network."
         exit ${SMF_EXIT_ERR_CONFIG}
